src/utils/video.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {once} from '../utils/function';
+import {setStyles} from '../style';
+
+
+/**
+ * @param {!Window} win
+ * @param {boolean} isLiteViewer
+ * @return {!Promise<boolean>}
+ */
+function isAutoplaySupportedImpl(win, isLiteViewer) {
+  // We do not support autoplay in amp-lite viewer regardless of platform.
+  if (isLiteViewer) {
+    return Promise.resolve(false);
+  }
+
+  // To detect autoplay, we create a video element and call play on it, if
+  // `paused` is true after `play()` call, autoplay is supported. Although
+  // this is unintuitive, it works across browsers and is currently the lightest
+  // way to detect autoplay without using a data source.
+  const detectionElement = win.document.createElement('video');
+
+  // NOTE(aghassemi): We need both attributes and properties due to Chrome and
+  // Safari differences when dealing with non-attached elements.
+  detectionElement.setAttribute('muted', '');
+  detectionElement.setAttribute('playsinline', '');
+  detectionElement.setAttribute('webkit-playsinline', '');
+  detectionElement.setAttribute('height', '0');
+  detectionElement.setAttribute('width', '0');
+
+  detectionElement.muted = true;
+  detectionElement.playsinline = true;
+  detectionElement.webkitPlaysinline = true;
+
+  setStyles(detectionElement, {
+    position: 'fixed',
+    top: '0',
+    width: '0',
+    height: '0',
+    opacity: '0',
+  });
+
+  new Promise(() => detectionElement.play()).catch(() => {
+    // Suppress any errors, useless to report as they are expected.
+  });
+
+  return Promise.resolve(!detectionElement.paused);
+}
+
+
+/**
+ * @param {!Window} win
+ * @param {boolean} isLiteViewer
+ * @return {!Promise<boolean>}
+ */
+let isAutoplaySupported = once(isAutoplaySupportedImpl);
```

Comment:
I think this assignment would be better if it was separated from the declaration and moved to inside `VideoUtils.isAutoplaySupported`.

Reply:
`once` returns a function, which is what the utility executes. Can't really assign that inside the class context unless it was instantiated, which would defeat the purpose.

Reply:
Hmm I'm don't know how it defeats the purpose. Can you help me understand? I know we generally discourage allocating memory at module-load time, and since `once` returns a new function with some closure variables it does allocate memory. I was thinking something like this:

```js
/** @type {?<insert-type-here>} */
let isAutoplaySupported = null;

/* ... */
export class VideoUtils {
  static isAutoplaySupported(win, isLiteViewer) {
    if (!isAutoplaySupported) {
      isAutoplaySupported = once(isAutoplaySupportedImpl);
    }
    return isAutoplaySupported(win, isLiteViewer);
  }
}
```

Reply:
Oh! I see what you mean.

I would argue that any function definition allocates memory. I like your approach though, changing.

Reply:
I think it might not only be memory allocation. My motivation for this suggestion is from an issue involving `map()` calls that run at module load time https://github.com/ampproject/amphtml/issues/6089. I could be cargo-culting my recommendation, but it does feel cleaner to not assign the result of a computation in the module-load context.

Reply:
sg, changed.